### ISSUE TITLE
"No such file or directory" is popped up for a successful "Find in Files"

### ISSUE THREAD

**Issue by sanjmeh** [rstudio/rstudio#12781]
### System details

  RStudio 2022.07.1+554 "Spotted Wakerobin" Release (7872775ebddc40635780ca1ed238934c3345c5de, 2022-07-22) for Ubuntu Bionic
Mozilla/5.0 (Macintosh; Intel Mac OS X 10.15; rv:109.0) Gecko/20100101 Firefox/110.0

R version:
platform       x86_64-pc-linux-gnu         
arch           x86_64                      
os             linux-gnu                   
system         x86_64, linux-gnu           
status                                     
major          4                           
minor          2.1                         
year           2022                        
month          06                          
day            23                          
svn rev        82513                       
language       R                           
version.string R version 4.2.1 (2022-06-23)
nickname       Funny-Looking Kid           
### Steps to reproduce the problem

The problem comes from time to time and this is not the first time. However I have no repeatable method figured out yet. If you give pointers i will try to replicate.

### Describe the problem in detail
When I search a string in all files by using `Shift + Cmd + F` the IDE is able to search the file and show a few search results in the pane (see screen grab) but on clicking on any of the search result it pops up this error window (see screen grab).

The error has started happening ever since I moved some files around.

Here is the screen grab of the error.
![image](https://user-images.githubusercontent.com/3091624/222465851-85809440-6fac-4918-9f8c-ea0b8c5895bb.png)

### Describe the behavior you expected
I expect it to open the file and take the cursor to the line containing the line - this is the usual behavior. 
But instead it is unable to even locate the file.

- [ x] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [ ] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [x ] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by sanjmeh** [comment]
I can confirm that the problem has now started in all the files. 100%.

**Comment by kevinushey** [comment]
If you can reproduce the issue again, please try the following steps after:

1. Use `Help -> Diagnostics -> Request Log` from the main menu;
2. Look for a bar with the name "open_document",
3. Copy the text you see at the bottom.

You should see something like the following there:

`````
Request ID: -1637814334

== REQUEST ======
{
  "method": "open_document",
  "params": [
    "~/r/pkg/renv/R/bioconductor.R",
    "r_source",
    "UTF-8"
  ],
  "clientId": "33e600bb-c1b1-46bf-b562-ab5cba070b0e",
  "clientVersion": ""
}

== RESPONSE ======
{
  "result": {
    "id": "E00A4F52",
    "path": "~/r/pkg/renv/R/bioconductor.R",
    "project_path": "R/bioconductor.R",
    "type": "r_source",
    "hash": "4036060993",
    "contents": "<...>",
    "dirty": false,
    "created": 1678052817413,
    "source_on_save": false,
    "relative_order": 1,
    "properties": {
      "source_window_id": "",
      "Source": "Source",
      "cursorPosition": "8,58",
      "scrollLine": "0",
      "marks": "<:1,0\n>:1,21"
    },
    "folds": "",
    "lastKnownWriteTime": 1678050433,
    "encoding": "UTF-8",
    "collab_server": "",
    "source_window": "",
    "last_content_update": 1678050433,
    "read_only": false,
    "read_only_alternatives": [],
    "extended_type": "",
    "notebook": {}
  },
  "ep": "false"
`````

except I expect your "response" text might include some error output.

**Comment by sanjmeh** [comment]
Got hold of the logs you desired. Used the `E` command (of rstudio) to export as a text file but that exported a lot of data. You want only the `open document` API response so here is the copy pasted section of the log that shows the error.
I will learn how to export only one API or group of API responses later.

```
""clientVersion"":""7872775ebddc40635780ca1ed238934c3345c5de""}",1,1678088892153,"{""result"":null,""ep"":""false""}"
1678088892766,1703630405,"{""method"":""stat"", ""params"":[""JSW-VTPL/oldfuncs/mariadb.R""], ""clientId"":""50952417-5f96-46a5-9021-6ee8bd5317fa"", ""clientVersion"":""7872775ebddc40635780ca1ed238934c3345c5de""}",1,1678088892876,"{""result"":{""path"":""~/JSW-VTPL/tablet/JSW-VTPL/oldfuncs/mariadb.R"",""raw_path"":""/home/sanjay/JSW-VTPL/tablet/JSW-VTPL/oldfuncs/mariadb.R"",""dir"":false,""length"":0,""exists"":false,""lastModified"":0.0},""ep"":""false""}"
1678088892893,-1194719203,"{""method"":""open_document"", ""params"":[""~/JSW-VTPL/tablet/JSW-VTPL/oldfuncs/mariadb.R"",""r_source"",""UTF-8""], ""clientId"":""50952417-5f96-46a5-9021-6ee8bd5317fa"", ""clientVersion"":""7872775ebddc40635780ca1ed238934c3345c5de""}",1,1678088892997,"{""error"":{""code"":100,""message"":""Error occurred while executing method"",""error"":{""code"":2,""category"":""system"",""message"":""No such file or directory"",""location"":""rstudio::core::Error rstudio::session::modules::source::{anonymous}::openDocument(const rstudio::core::json::JsonRpcRequest&, rstudio::core::json::JsonRpcResponse*) src/cpp/session/modules/SessionSource.cpp:319""}}}"
1678088892998,-1200010049,"{""method"":""list_remove_item"", ""params"":[""file_mru"",""~/JSW-VTPL/tablet/JSW-VTPL/oldfuncs/mariadb.R""], ""clientId"":""50952417-5f96-46a5-9021-6ee8bd5317fa"", ""clientVersion"":""7872775ebddc40635780ca1ed238934c3345c5de""}",1,1678088893069,"{""result"":null,""ep"":""false""}"
1678088893185,-2120652764,"{""method"":""get_events"", ""params"":[122645], ""clientId"":""50952417-5f96-46a5-9021-6ee8bd5317fa"", ""clientVersion"":""7872775ebddc40635780ca1ed238934c3345c5de""}",1,1678088895723,"{""result"":[{""id"":122646,""type"":""memory_usage_changed"",""data"":{""total"":{""kb"":16392556,""provider"":6},""used"":{""kb"":5634116,""provider"":6},""process"":{""kb"":1097112,""provider"":5}}}],""ep"":""false""}"
1678088895725,-377314803,"{""method"":""get_events"", ""params"":[122646], ""clientId"":""50952417-5f96-46a5-9021-6ee8bd5317fa"", ""clientVersion"":""7872775ebddc40635780ca1ed238934c3345c5de""}",1,1678088905559,"{""result"":[{""id"":122647,""type"":""memory_usage_changed"",""data"":{""total"":{""kb"":16392556,""provider"":6},""used"":{""kb"":5711616,""provider"":6},""process"":{""kb"":1097112,""provider"":5}}}],""ep"":""false""}"
```

**Comment by sanjmeh** [comment]
> 



> @sanjmeh Thank you for the logs! @kevinushey can interpret the logs far, far better than I could, but are you by chance using a shared drive?

No, I have a dedicated Ubuntu server with a dedicated root volume and the paths are all accessible through Rstudio files tab, so the drive access cannot be an issue. It is most likely a relative path problem.

**Comment by kevinushey** [comment]
Thanks -- including the relevant output, with some nicer formatting:

```
{
  "method": "open_document",
  "params": [
    "~/JSW-VTPL/tablet/JSW-VTPL/oldfuncs/mariadb.R",
    "r_source",
    "UTF-8"
  ],
  "clientId": "50952417-5f96-46a5-9021-6ee8bd5317fa",
  "clientVersion": "7872775ebddc40635780ca1ed238934c3345c5de"
}
{
  "error": {
    "code": 100,
    "message": "Error occurred while executing method",
    "error": {
      "code": 2,
      "category": "system",
      "message": "No such file or directory",
      "location": "rstudio::core::Error rstudio::session::modules::source::{anonymous}::openDocument(const rstudio::core::json::JsonRpcRequest&, rstudio::core::json::JsonRpcResponse*) src/cpp/session/modules/SessionSource.cpp:319"
    }
  }
}
```

Does this path ("~/JSW-VTPL/tablet/JSW-VTPL/oldfuncs/mariadb.R") look correct?

The most likely explanation is there we're somehow resolving the home directory incorrectly here. Can you also share the output of:

```
path.expand("~/")
normalizePath("~/")
Sys.getenv("HOME")
```

I wonder if those are out-of-sync somehow?

**Comment by sanjmeh** [comment]
> Does this path ("~/JSW-VTPL/tablet/JSW-VTPL/oldfuncs/mariadb.R") look correct?

NO! This is not correct. 

The correct path is  `~/JSW-VTPL/oldfuncs/mariadb.R`
and the directory `~/JSW-VTPL/tablet` also exists and is usually my current working directory.


> path.expand("~/")
> normalizePath("~/")
> Sys.getenv("HOME")

And all the above correctly return my home directory
`/home/sanjay`

So for some reason the path `JSW-VTPL/oldfuncs` is getting added. 

I can also tell you I was trying to create a package and then reverted to normal functions. That may have caused some path variables tinkering but not sure why are they still haunting me when I am not writing a package now.

**Comment by ronblum** [comment]
@kevinushey I can reproduce this in RStudio Server (Workbench version) 2023.03.0+385.pro1 when not working in a project. When working in a project, find in files works.

**Comment by ronblum** [comment]
I can no longer reproduce the issue (#12805). 🤔

**Comment by kevinushey** [comment]
> I can also tell you I was trying to create a package and then reverted to normal functions. That may have caused some path variables tinkering but not sure why are they still haunting me when I am not writing a package now.

Are you working within an RStudio project? If so, do you see this same sort of behavior within other projects? I'm still at a bit of a loss as to what's going on here.

**Comment by sanjmeh** [comment]
> Are you working within an RStudio project? If so, do you see this same sort of behavior within other projects? I'm still at a bit of a loss as to what's going on here.

No I am not working under any Rstudio Project. The above comment was because one of my colleagues defined a package structure as I have a lot of functions being re used across projects but later reverted it, as it was causing path problems.. not this but application path problems. However we did test all the paths you mentioned and they all pointed to my home.

**Comment by sanjmeh** [comment]
I can see the problem now everytime in the log, reproducing it here:
```
Request ID: 611578136

== REQUEST ======
{
  "method": "open_document",
  "params": [
    "~/JSW-VTPL/omni/mobifox/api.mfox.R",
    "r_source",
    "UTF-8"
  ],
  "clientId": "7b9583d5-1f95-4c70-ac28-7953db589c11",
  "clientVersion": "7872775ebddc40635780ca1ed238934c3345c5de"
}

== RESPONSE ======
{
  "error": {
    "code": 100,
    "message": "Error occurred while executing method",
    "error": {
      "code": 2,
      "category": "system",
      "message": "No such file or directory",
      "location": "rstudio::core::Error rstudio::session::modules::source::{anonymous}::openDocument(const rstudio::core::json::JsonRpcRequest&, rstudio::core::json::JsonRpcResponse*) src/cpp/session/modules/SessionSource.cpp:319"
    }
  }
}
```

If I see the path it is trying to open:
` ~/JSW-VTPL/omni/mobifox/api.mfox.R`

**Comment by kevinushey** [comment]
Can you please also confirm some of the output you see in the console if you run

```
Sys.setenv(RSTUDIO_GREP_DEBUG = 1)
```

before running Find in Files?

**Comment by sanjmeh** [comment]
> Can you please also confirm some of the output you see in the console if you run

Here you go..

```
grep '--binary-files=without-match' '-rHn' '--color=always' '--devices=skip' '-i' '-F' '-f' '/tmp/RtmpfrxsA4/rs_grep3253e06645f32.txt' '--include=*.r' '--include=*.R'
stdout: omni/DBL/api.dbl.R:23:source("../commonAPI.R")
omni/mobifox/api.mfox.R:1:# API functions tailored for ABL. Prefer these over commonAPI.R
omni/abl/api.abl.R:1:# API functions tailored for ABL. Prefer these over commonAPI.R
omni/abl/email.R:4:suppressMessages(source("~/omni/commonAPI.R"))
omni/airtel/omniapi.R:2:if(!exists("loaded.commonapi.r"))
omni/airtel/omniapi.R:3:  source("~/omni/commonAPI.R")
omni/inchara/api.inchara.R:24:source("~/omni/commonAPI.R")
omni/cron_statistics.R:3:source("~/omni/commonAPI.R")
omni/smallfns.R:359:  remove("loaded.commonapi.r",envir = .GlobalEnv)
omni/projdrain.R:6:source("~/omni/commonAPI.R")
omni/apco/api.apco.R:26:source("~/omni/commonAPI.R")
omni/commonAPI.R:3:loaded.commonapi.r <- T
omni/commonAPI.R:294:# this is in commonAPI.R; we have to remove the individual get_events() in client directory
omni/cron_functions.R:3:if(!exists("loaded.commonapi.r")) source("~/omni/commonAPI.R")
omni/cron_functions.R:9:#source("~/omni/airtel/omniapi.R") # for airtel only - we have to get rid of omniapi.R and move completely to commonAPI.R
JSW-VTPL/oldfuncs/cron_functions.R:4:if(!exists("loaded.commonapi.r")) source("~/omni/commonAPI.R")
JSW-VTPL/oldfuncs/cron_functions.R:10:#source("~/omni/airtel/omniapi.R") # for airtel only - we have to get rid of omniapi.R and move completely to commonAPI.R
JSW-VTPL/jswapi.R:295:# this is in commonAPI.R; we have to remove the individual get_events() in client directory
gpt/smallfns.R:338:  remove("loaded.commonapi.r",envir = .GlobalEnv)
TMA/api.R:257:# this is in commonAPI.R; we have to remove the individual get_events() in client directory

```

**Comment by kevinushey** [comment]
Thanks! I think this confirms that RStudio is erroneously placing `~/JSW-VTPL` rather than `~/` in front of these paths, but it's still not clear to me why.

Is there any chance that that `HOME` environment variable (as visible to RStudio in some location) was modified at some point in time?

Or, perhaps, is the `R_USER` environment variable set to something?

**Comment by kevinushey** [comment]
Yes please, that may be helpful.

**Comment by SimonDedman** [comment]
FWIW this has been happening to me seemingly every time I open RStudio, switch projects, or open a script. Can send diagnostics if required but maybe best to wait to see the outcome of this exchange.

**Comment by sanjmeh** [comment]
> Shall I paste the whole `env`output on shell?

Here you go..
```
anjay@rstudio-blr-01:/data$ env
RSTUDIO_CLI_HYPERLINKS=true
RSTUDIO_TERM=DE8C2E81
R_BZIPCMD=/bin/bzip2
R_STRIP_STATIC_LIB=strip --strip-debug
HISTCONTROL=ignoreboth
R_CLI_HAS_HYPERLINK_IDE_HELP=true
R_GZIPCMD=/bin/gzip -n
R_HOME=/usr/lib/R
R_CLI_HAS_HYPERLINK_IDE_RUN=true
RSTUDIO_USER_IDENTITY=sanjay
RSTUDIO_SESSION_STREAM=sanjay-d
RS_RPOSTBACK_PATH=/usr/lib/rstudio-server/bin/rpostback
_R_CHECK_COMPILATION_FLAGS_KNOWN_=-Wformat -Werror=format-security -Wdate-time
RS_SESSION_TMP_DIR=/var/run/rstudio-server/rstudio-rsession
RS_SERVER_RPC_SOCKET_PATH=/var/run/rstudio-server/rstudio-rserver/session-server-rpc.socket
SED=/bin/sed
RSTUDIO_R_MODULE=
RMARKDOWN_MATHJAX_PATH=/usr/lib/rstudio-server/resources/mathjax-27
TAR=/bin/tar
R_CLI_HAS_HYPERLINK_IDE_VIGNETTE=true
EDITOR=vi
RSTUDIO_SESSION_PID=922287
PWD=/data
LOGNAME=sanjay
RSTUDIO_PROJ_NAME=JSW.VTPL
RSTUDIO_R_VERSION_LABEL=
R_LIBS_SITE=/usr/local/lib/R/site-library/:/usr/lib/R/site-library:/usr/lib/R/library
R_PLATFORM=x86_64-pc-linux-gnu
RSTUDIO_WINUTILS=bin/winutils
R_TEXI2DVICMD=/usr/bin/texi2dvi
RSTUDIO_SESSION_ID=46912c1b
RSTUDIO_R_REPO=
HOME=/home/sanjay
R_RD4PDF=times,inconsolata,hyper
SSH_ASKPASS=rpostback-askpass
LANG=C.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
RSTUDIO_CONSOLE_COLOR=256
MPLENGINE=tkAgg
R_SYSTEM_ABI=linux,gcc,gxx,gfortran,gfortran
RSTUDIO_PROGRAM_MODE=server
GIT_ASKPASS=rpostback-askpass
PROMPT_COMMAND=echo -ne "\033]0;${PWD/#${HOME}/~}\007"
RSTUDIO_USER_IDENTITY_DISPLAY=sanjay
R_INCLUDE_DIR=/usr/share/R/include
CLICOLOR_FORCE=1
R_LIBS_USER=/home/sanjay/R/x86_64-pc-linux-gnu-library/4.2
R_PAPERSIZE=letter
R_UNZIPCMD=/usr/bin/unzip
R_BROWSER=xdg-open
LESSCLOSE=/usr/bin/lesspipe %s %s
TERM=xterm-256color
R_STRIP_SHARED_LIB=strip --strip-unneeded
R_PAPERSIZE_USER=a4
LESSOPEN=| /usr/bin/lesspipe %s
USER=sanjay
R_PDFVIEWER=/usr/bin/xdg-open
R_DOC_DIR=/usr/share/R/doc
TZDIR=/usr/share/zoneinfo
R_PRINTCMD=/usr/bin/lpr
DISPLAY=:0
SHLVL=58
R_SESSION_TMPDIR=/tmp/RtmpYVmyTv
PAGER=/usr/bin/pager
RSTUDIO_CONSOLE_WIDTH=161
RSTUDIO=1
LD_LIBRARY_PATH=/usr/lib/R/lib:/lib:/usr/lib/x86_64-linux-gnu:/usr/lib/jvm/default-java/lib/server
RSTUDIO_R_PRELAUNCH_SCRIPT=
RSTUDIO_CHILD_PROCESS_PANE=terminal
R_ZIPCMD=/usr/bin/zip
RSTUDIO_HTTP_REFERER=http://64.227.177.82:8787/
MAKE=make
R_SHARE_DIR=/usr/share/R/share
XDG_DATA_DIRS=/usr/local/share:/usr/share:/var/lib/snapd/desktop
PATH=/home/sanjay/.local/bin:/home/sanjay/.local/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin:/usr/lib/rstudio-server/bin/postback
RSTUDIO_PANDOC=/usr/lib/rstudio-server/bin/quarto/bin/tools
LN_S=ln -s
OLDPWD=/home/sanjay/JSW-VTPL/tablet
_=/usr/bin/env
```

**Comment by kevinushey** [comment]
> OLDPWD=/home/sanjay/JSW-VTPL/tablet

This looks suspicious to me. Is it possible that your `HOME` directory was temporarily set to this location, and RStudio Server was started using that as your `HOME`? If so, RStudio probably is "remembering" that location and using that when resolving paths for Find in Files.

**Comment by sanjmeh** [comment]
Hmm..
A quick echo of $HOME always returns the correct home. Is it possible that bash is saying something as HOME while rstudio something else? How does Rstudio decide what is HOME? 
I also checked by
`list.files("~")` and the files listed are correctly my home directory file.
So it seems the expand file name script of rstudio has some weird bug.

**Comment by mikebessuille** [comment]
@sanjmeh can you please try the Elsbeth Geranium version, 2022.12.0?  We fixed some related problems and I'm hoping this problem is also fixed.

**Comment by sanjmeh** [comment]
Sorry this is a little urgent as I have removed the old rstudio already:
Earlier rstudio was:
`RStudio 2022.07.1+554 "Spotted Wakerobin" Release (7872775ebddc40635780ca1ed238934c3345c5de, 2022-07-22) for Ubuntu Bionic` and now I am confused about which version and from where do i get it.

Now I am ready to install from here but this does not look to be the latest release.
`wget https://download2.rstudio.org/server/bionic/amd64/rstudio-server-2022.12.0-353-amd64.deb`
Is this the Elsbeth Germanium version, and should I proceed?
